File: android/src/org/coolreader/crengine/TOCDlg.java
/*
 * CoolReader for Android
 * Copyright (C) 2010-2012 Vadim Lopatin <coolreader.org@gmail.com>
 *
 * This program is free software: you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation, either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

package org.coolreader.crengine;

import android.database.DataSetObserver;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;

import org.coolreader.R;

import java.util.ArrayList;

public class TOCDlg extends BaseDialog {
	//CoolReader mCoolReader;
	ReaderView mReaderView;
	TOCItem mTOC;
	ListView mListView;
	int mCurrentPage;
	TOCItem mCurrentPageItem;
	ArrayList<TOCItem> mItems = new ArrayList<>();
	private LayoutInflater mInflater;
	
	private void initItems( TOCItem toc, boolean expanded )
	{
		for ( int i=0; i<toc.getChildCount(); i++ ) {
			TOCItem child = toc.getChild(i);
			if ( child.getPage()<=mCurrentPage )
				mCurrentPageItem = child;
			if ( expanded ) {
				child.setGlobalIndex(mItems.size());
				mItems.add(child);
			} else {
				child.setGlobalIndex(-1); // invisible
			}
			initItems(child, expanded && child.getExpanded());
		}
	}
	private void initItems()
	{
		mCurrentPageItem = null;
		mItems.clear();
		initItems(mTOC, true);
	}
	
	private void expand( TOCItem item )
	{
		if ( item==null )
			return;
		item.setExpanded(true);
		// expand all parents
		for ( TOCItem p = item.getParent(); p!=null; p = p.getParent() )
			p.setExpanded(true);
		initItems();
		refreshList();
		if ( mItems.size()>0 ) {
			if ( item.getGlobalIndex()>=0 ) {
				mListView.setSelection(item.getGlobalIndex());
				mListView.setSelectionFromTop(item.getGlobalIndex(), mListView.getHeight()/2);
			} else
				mListView.setSelection(0);
		}
	}
	
	private void collapse( TOCItem item )
	{
		item.setExpanded(false);
		initItems();
		refreshList();
	}
	
	private void refreshList()
	{
		mListView.setAdapter(new BaseAdapter() {
			public boolean areAllItemsEnabled() {
				return true;
			}

			public boolean isEnabled(int arg0) {
				return true;
			}

			public int getCount() {
				return mItems.size();
			}

			public Object getItem(int position) {
				return mItems.get(position);
			}

			public long getItemId(int position) {
				return position;
			}

			public int getItemViewType(int position) {
				TOCItem item = mItems.get(position);
				boolean isCurrentItem = item==mCurrentPageItem;
				return isCurrentItem ? 0 : 1;
			}

			
			public View getView(int position, View convertView, ViewGroup parent) {
				View view;
				TOCItem item = mItems.get(position);
				boolean isCurrentItem = item==mCurrentPageItem;
				if ( convertView==null ) {
					//view = new TextView(getContext());
					view = mInflater.inflate(isCurrentItem ? R.layout.toc_item_current : R.layout.toc_item, null);
				} else {
					view = convertView;
				}
				TextView pageTextView = view.findViewById(R.id.toc_page);
				TextView titleTextView = view.findViewById(R.id.toc_title);
				TextView marginTextView = view.findViewById(R.id.toc_level_margin);
				ImageView expandImageView = view.findViewById(R.id.toc_expand_icon);
				StringBuilder buf = new StringBuilder(item.getLevel()*2);
				for ( int i=1; i<item.getLevel(); i++ )
					buf.append("  ");
				if ( item.getChildCount()>0 ) {
					if ( item.getExpanded() ) {
						expandImageView.setImageResource(R.drawable.cr3_toc_item_expanded);
					} else {
						expandImageView.setImageResource(R.drawable.cr3_toc_item_collapsed);
					}
				} else {
					expandImageView.setImageResource(R.drawable.cr3_toc_item_normal);
				}
				marginTextView.setText(buf.toString());
				titleTextView.setText(item.getName());
				pageTextView.setText(String.valueOf(item.getPage()+1));
				return view;
			}

			public int getViewTypeCount() {
				return 2;
			}

			public boolean hasStableIds() {
				return true;
			}

			public boolean isEmpty() {
				return false;
			}

			private ArrayList<DataSetObserver> observers = new ArrayList<>();
			
			public void registerDataSetObserver(DataSetObserver observer) {
				observers.add(observer);
			}

			public void unregisterDataSetObserver(DataSetObserver observer) {
				observers.remove(observer);
			}
		});
	}

	public TOCDlg(BaseActivity coolReader, ReaderView readerView, TOCItem toc, int currentPage )
	{
		super(coolReader, coolReader.getResources().getString(R.string.win_title_toc), false, false);
        setCancelable(true);
//		this.mCoolReader = coolReader;
		this.mReaderView = readerView;
		this.mTOC = toc;
		this.mCurrentPage = currentPage;
		this.mListView = new BaseListView(getContext(), true);
		mListView.setOnItemClickListener((listview, view, position, id) -> {
			TOCItem item = mItems.get(position);
			if ( item.getChildCount()==0 || item.getExpanded() ) {
				mReaderView.goToPage(item.getPage()+1);
				dismiss();
			} else {
				expand(item);
			}
		});
		mListView.setOnItemLongClickListener((listview, view, position, id) -> {
			TOCItem item = mItems.get(position);
			if ( item.getChildCount()==0 ) {
				mReaderView.goToPage(item.getPage()+1);
				dismiss();
			} else {
				if ( item.getExpanded() )
					collapse(item);
				else
					expand(item);
			}
			return true;
		});
		mListView.setLongClickable(true);
		mListView.setClickable(true);
		mListView.setFocusable(true);
		mListView.setFocusableInTouchMode(true);
		mListView.setChoiceMode(ListView.CHOICE_MODE_SINGLE);
		setView(mListView);
		// cancel
		//
		setFlingHandlers(mListView, TOCDlg.this::dismiss, TOCDlg.this::dismiss);
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
        mInflater = LayoutInflater.from(getContext());
		super.onCreate(savedInstanceState);
		expand( mTOC );
		expand( mCurrentPageItem );
	}
}

File: cr3qt/src/tocdlg.cpp
/***************************************************************************
 *   CoolReader, Qt GUI                                                    *
 *   Copyright (C) 2009-2011,2014 Vadim Lopatin <coolreader.org@gmail.com> *
 *   Copyright (C) 2018,2020 Aleksey Chernov <valexlin@gmail.com>          *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or         *
 *   modify it under the terms of the GNU General Public License           *
 *   as published by the Free Software Foundation; either version 2        *
 *   of the License, or (at your option) any later version.                *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the Free Software           *
 *   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,            *
 *   MA 02110-1301, USA.                                                   *
 ***************************************************************************/

#include "tocdlg.h"
#include "ui_tocdlg.h"
#include "cr3widget.h"
#include "../crengine/include/lvdocview.h"
#include <QStandardItemModel>
#include "crqtutil.h"

class TocItem : public QTreeWidgetItem
{
        LVTocItem * _item;
    public:
        LVTocItem * getItem() { return _item; }
        TocItem( LVTocItem * item, int currPage, int & nearestPage, TocItem * & nearestItem )
                : QTreeWidgetItem( QStringList()
                                   << (item ? cr2qt( item->getName() ) : "No TOC items" )
                                   << ( item ? cr2qt(lString32::itoa(item->getPage()+1)) : "")
                                   )
                , _item( item )
        {
            if (item) {
                int page = item->getPage();
                if ( !nearestItem || (page <= currPage && page > nearestPage) ) {
                    nearestItem = this;
                    nearestPage = page;
                }

                setData( 0, Qt::UserRole, QVariant( cr2qt(item->getPath()) ) );
                for ( int i=0; i<item->getChildCount(); i++ ) {
                    addChild( new TocItem( item->getChild(i), currPage, nearestPage, nearestItem ) );
                }
            }
        }
};

#if QT_VERSION >= 0x050000
#define setResizeModeMethod setSectionResizeMode
#else
#define setResizeModeMethod setResizeMode
#endif

bool TocDlg::showDlg(  QWidget * parent, CR3View * docView )
{
    LVTocItem * root = docView->getToc();
    if ( !root || !root->getChildCount() )
        return false;
    TocDlg * dlg = new TocDlg( parent, docView );
    dlg->show();
    return true;
}

TocDlg::TocDlg(QWidget *parent, CR3View * docView) :
    QDialog(parent),
    m_ui(new Ui::TocDlg), m_docview(docView)
{
    setAttribute( Qt::WA_DeleteOnClose, true );
    m_ui->setupUi(this);
    m_ui->treeWidget->setColumnCount(2);
    m_ui->treeWidget->setHeaderItem(new QTreeWidgetItem(QStringList() << tr("Title") << tr("Page") ));
    m_ui->treeWidget->header()->setStretchLastSection(false);
    m_ui->treeWidget->header()->setResizeModeMethod( 0, QHeaderView::Stretch );
    m_ui->treeWidget->header()->setResizeModeMethod( 1, QHeaderView::ResizeToContents );

    int nearestPage = -1;
    int currPage = docView->getCurPage();
    TocItem * nearestItem = NULL;
    LVTocItem * root = m_docview->getToc();
    for ( int i=0; i<root->getChildCount(); i++ ) {
        m_ui->treeWidget->addTopLevelItem( new TocItem( root->getChild(i), currPage, nearestPage, nearestItem ) );
    }
    m_ui->treeWidget->expandAll();
    if ( nearestItem ) {
        m_ui->treeWidget->setCurrentItem( nearestItem );
    }
    m_docview->restoreWindowPos( this, "toc." );
}

void TocDlg::closeEvent ( QCloseEvent * event )
{
    m_docview->saveWindowPos( this, "toc." );
}

TocDlg::~TocDlg()
{
    delete m_ui;
}

void TocDlg::changeEvent(QEvent *e)
{
    switch (e->type()) {
    case QEvent::LanguageChange:
        m_ui->retranslateUi(this);
        break;
    default:
        break;
    }
}

void TocDlg::on_treeWidget_doubleClicked(QModelIndex index)
{
    QString s = index.data(Qt::UserRole).toString();
    m_docview->goToXPointer(s);
    close();
}

void TocDlg::on_buttonBox_accepted()
{
    on_treeWidget_doubleClicked(m_ui->treeWidget->currentIndex());
}

void TocDlg::on_buttonBox_rejected()
{
    close();
}
